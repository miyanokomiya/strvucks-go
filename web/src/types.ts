export interface User {
  id: number
  athleteId: number
  username: string
  iftttKey: string
  iftttMessage: string
}

export interface Summary {
  id: number
  athleteId: number

  latestDistance: number
  latestMovingTime: number
  latestTotalElevationGain: number
  latestCalories: number

  monthBaseDate: string
  monthlyCount: number
  monthlyDistance: number
  monthlyMovingTime: number
  monthlyTotalElevationGain: number
  monthlyCalories: number

  weekBaseDate: string
  weeklyCount: number
  weeklyDistance: number
  weeklyMovingTime: number
  weeklyTotalElevationGain: number
  weeklyCalories: number
}

