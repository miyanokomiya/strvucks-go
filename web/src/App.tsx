import * as React from 'react'
import Link from '@material-ui/core/Link'
import axios from 'axios'
import queryString from 'query-string'
import Grid from '@material-ui/core/Grid'
import Button from '@material-ui/core/Button'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import Snackbar from '@material-ui/core/Snackbar'
import IconButton from '@material-ui/core/IconButton'
import CloseIcon from '@material-ui/icons/Close'
import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import { User, Summary } from './types'
import { SummaryCard } from './components/Summary'

const parsed = queryString.parse(location.search)
const token = (parsed.token as string) || localStorage.getItem('token') || ''
axios.defaults.headers.common['Authorization'] = token

export const App: React.FC = () => {
  const [loading, setLoading] = React.useState(true)
  const [user, setUser] = React.useState(null as User | null)
  const [summary, setSummary] = React.useState(null as Summary | null)
  const [stravaAuth, setStravaAuth] = React.useState('')
  const [draftKey, setDraftKey] = React.useState('')
  const [draftMessage, setDraftMessage] = React.useState('')
  const [snack, setSnack] = React.useState('')

  const onCloseSnack = React.useCallback(() => {
    setSnack('')
  }, [])
  const onInputDraftKey = React.useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    setDraftKey(e.currentTarget.value)
  }, [])
  const onInputDraftMessage = React.useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    setDraftMessage(e.currentTarget.value)
  }, [])
  const onSubmit = React.useCallback(
    (e: React.FormEvent<HTMLFormElement>) => {
      e.preventDefault()
      setLoading(true)
      axios
        .post('/api/current_user', {
          iftttKey: draftKey,
          iftttMessage: draftMessage,
        })
        .then(res => {
          setUser(res.data)
          setSnack('Saved')
          setLoading(false)
        })
        .catch(() => {
          setLoading(false)
        })
    },
    [draftKey, draftMessage],
  )

  const onRecalc = React.useCallback(() => {
    setLoading(true)
    axios
      .put('/api/current_user/recalc')
      .then(res => {
        setSummary(res.data)
        setSnack('Recalced')
        setLoading(false)
      })
      .catch(() => {
        setLoading(false)
      })
  }, [])

  React.useEffect(() => {
    axios
      .get('/api/current_user')
      .then(res => {
        setUser(res.data)
        setDraftKey(res.data.iftttKey)
        setDraftMessage(res.data.iftttMessage)
        localStorage.setItem('token', token)
        setLoading(false)
      })
      .catch(() => {
        setLoading(false)
      })
  }, [])

  React.useEffect(() => {
    axios.get('/api/current_user/summary').then(res => {
      setSummary(res.data)
    })
  }, [])

  React.useEffect(() => {
    axios.get('/api/strava_auth').then(res => {
      setStravaAuth(res.data.url)
    })
  }, [])

  const userBlock = React.useMemo(() => {
    if (user)
      return (
        <form onSubmit={onSubmit}>
          <Grid container>
            <Grid item xs={12}>
              <Typography>ID: {user.id}</Typography>
            </Grid>
            <Grid item xs={12}>
              <Typography>Strava ID: {user.athleteId}</Typography>
            </Grid>
            <Grid item xs={12}>
              <Typography>Strava Username: {user.username}</Typography>
            </Grid>
            <Grid item xs={12}>
              <TextField
                label="IFTTT Key"
                variant="outlined"
                margin="normal"
                value={draftKey}
                onChange={onInputDraftKey}
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                label="IFTTT Message"
                variant="outlined"
                margin="normal"
                value={draftMessage}
                onChange={onInputDraftMessage}
              />
            </Grid>
            <Grid item>
              <Button type="submit" variant="contained" color="primary">
                Submit
              </Button>
            </Grid>
          </Grid>
          <Snackbar
            anchorOrigin={{
              vertical: 'bottom',
              horizontal: 'left',
            }}
            open={!!snack}
            autoHideDuration={6000}
            onClose={onCloseSnack}
            ContentProps={{
              'aria-describedby': 'message-id',
            }}
            message={<span>{snack}</span>}
            action={[
              <IconButton key="close" aria-label="close" color="inherit" onClick={onCloseSnack}>
                <CloseIcon />
              </IconButton>,
            ]}
          />
        </form>
      )
    return (
      <div>
        <Link href={stravaAuth}>
          <p>Login by Strava</p>
          <img src="/assets/strava.jpg" style={{ width: '120px', height: 'auto' }} />
        </Link>
      </div>
    )
  }, [
    draftKey,
    draftMessage,
    onCloseSnack,
    onInputDraftKey,
    onInputDraftMessage,
    onSubmit,
    snack,
    stravaAuth,
    user,
  ])

  const summaryBlock = React.useMemo(() => {
    if (summary)
      return (
        <>
          <SummaryCard summary={summary} />
          <Button type="button" variant="contained" color="primary" onClick={onRecalc}>
            Recalc
          </Button>
        </>
      )

    return (
      <Card>
        <CardContent>
          <Typography variant="h5" component="h2">
            No Summary
          </Typography>
        </CardContent>
      </Card>
    )
  }, [onRecalc, summary])

  const mainBlock = React.useMemo(() => {
    if (!loading)
      return (
        <div>
          <div>{userBlock}</div>
          <div>{summaryBlock}</div>
        </div>
      )

    return <div>loading...</div>
  }, [loading, summaryBlock, userBlock])

  return <div>{mainBlock}</div>
}
