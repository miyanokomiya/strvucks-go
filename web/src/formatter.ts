const formatter = new Intl.NumberFormat('ja', {
  useGrouping: true,
  minimumFractionDigits: 0,
  maximumFractionDigits: 0,
})

export const formatNatural = (num: number): string => {
  return formatter.format(num)
}

export const formatTime = (sec: number): string => {
  const h = Math.floor(sec / 60 / 60)
  const m = Math.floor(sec / 60 - 60 * h)
  const s = sec % 60
  if (h === 0) {
    return `${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`
  }
  return `${h}:${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`
}

export const formatLap = (meter: number, sec: number): string => {
  const lap = sec / (meter / 1000)
  const m = Math.floor(lap / 60)
  const s = Math.floor(lap - 60 * m)
  return `${m}:${String(s).padStart(2, '0')}/km`
}
