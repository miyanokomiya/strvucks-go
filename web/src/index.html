<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>strvucks</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="text/javascript" src="./index.tsx"></script>
  </body>
</html>
