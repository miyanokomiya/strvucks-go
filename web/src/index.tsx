import 'regenerator-runtime/runtime'
import * as React from 'react'
import { render } from 'react-dom'
import { App } from './App'

render(<App />, document.getElementById('root') as any)
