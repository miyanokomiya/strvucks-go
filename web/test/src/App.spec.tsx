import * as React from 'react'
import axios from 'axios'
import renderer from 'react-test-renderer'
import { actWait } from '../helper'
import { App } from '../../src/App'
import { mockUser, mockSummary } from '../mock'

jest.mock('axios')

describe('src/App', () => {
  describe('snapshot', () => {
    ;(axios.get as jest.Mock).mockImplementation((url: string) => {
      if (url == '/api/current_user')
        return Promise.resolve({
          data: mockUser(),
        })
      if (url == '/api/current_user/summary')
        return Promise.resolve({
          data: mockSummary(),
        })
      if (url == '/api/strava_auth')
        return Promise.resolve({
          data: { url: 'strava_auth_url' },
        })
      return Promise.reject(new Error('error: ' + url))
    })

    it('loading', () => {
      const tree = renderer.create(<App />).toJSON()
      expect(tree).toMatchSnapshot()
    })
    it('loaded', async () => {
      const root = renderer.create(<App />)
      await actWait()
      expect(root.toJSON()).toMatchSnapshot()
    })
  })
})
