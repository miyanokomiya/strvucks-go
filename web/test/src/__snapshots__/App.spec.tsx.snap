// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/App snapshot loaded 1`] = `
<div>
  <div>
    <div>
      <form
        onSubmit={[Function]}
      >
        <div
          className="MuiGrid-root MuiGrid-container"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              ID: 
              1
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              Strava ID: 
              10
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              Strava Username: 
              my name
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink={false}
              >
                IFTTT Key
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="text"
                  value="my key"
                />
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-192 MuiOutlinedInput-notchedOutline"
                  style={
                    Object {
                      "paddingLeft": 8,
                    }
                  }
                >
                  <legend
                    className="PrivateNotchedOutline-legend-193"
                    style={
                      Object {
                        "width": 0.01,
                      }
                    }
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink={false}
              >
                IFTTT Message
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="text"
                  value="my message"
                />
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-192 MuiOutlinedInput-notchedOutline"
                  style={
                    Object {
                      "paddingLeft": 8,
                    }
                  }
                >
                  <legend
                    className="PrivateNotchedOutline-legend-193"
                    style={
                      Object {
                        "width": 0.01,
                      }
                    }
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              disabled={false}
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="submit"
            >
              <span
                className="MuiButton-label"
              >
                Submit
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </form>
    </div>
    <div>
      <div
        className="MuiPaper-root MuiPaper-elevation1 MuiCard-root MuiPaper-rounded"
      >
        <div
          className="MuiCardContent-root"
        >
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            Activity Summary
          </h6>
          <h6
            className="MuiTypography-root MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
          >
            Latest
          </h6>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            3
             m
          </p>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            00m
          </p>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            22:13/km
          </p>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            6
             kcal
          </p>
          <h6
            className="MuiTypography-root MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
          >
            Weekly (
            1004
            )
          </h6>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            1,005
             m
          </p>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            27h46m
          </p>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            1658:28/km
          </p>
          <h6
            className="MuiTypography-root MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
          >
            Monthly (
            14
            )
          </h6>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            14
             m
          </p>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            00m
          </p>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            17:51/km
          </p>
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Recalc
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`src/App snapshot loading 1`] = `
<div>
  <div>
    loading...
  </div>
</div>
`;
